### ISSUE TITLE
NaN loss when using half precision

### ISSUE THREAD

**Issue by alexge233** [pytorch/pytorch#8860]
Let me start by saying I've googled the issue and understand others have also run into it.
I've tried with Adam optimizer and by setting epsilon to `1e-4` and `1e-3` but after a few iterations my loss is still `NaN`.
Because I'm looking at very large datasets and `float` to `half` does make a difference, is there any other trick I can try in order to keep using `half`? 
One of the posts on the forum suggests (https://discuss.pytorch.org/t/adam-half-precision-nans/1765) to keep a copy of the weights in `float32` but it is unclear as to how to do that?
The network is a copy of HappyNet, with 5 convolutional layers, 2 fully connected one, and a softmax output.

**Comment by alexge233** [comment]
Nevermind I managed to solve it by doing the following:
- setting `eps=1e-4`
- clamping the `Conv2d` layer propagation with `clamp(min=1e-4)`

**Comment by PyxAI** [comment]
This worked for me only with changing eps=1e-4
thanks!

**Comment by bluesky314** [comment]
How did you apply clamping to all the Conv2d layers? Is it done at every forward pass?

**Comment by alexge233** [comment]
@bluesky314 you define it in your `forward` method of your network.
